### PR TITLE
allows container on child objects

### LINKED ISSUE #841: Can't put containers on child game object when the parent game object is networked
## Summary

When trying to add a container on the maintenance panel on a vendor machine, a network identity is automatically created on the maintenance panel because some scripts necessary for the container to work are networkBehaviour. 
Due to some limitation with Mirror, there can't be two network identity in a game object hierarchy. 
This make impossible adding container to a child game object when the parent game object is networked, which is the case of
pretty much all game objects.

image

## Possible Solution

Adding the scripts with a networkbehaviour on the parent game object instead of on the child game object should solve it.
Since containerDescriptor is the only script a user is supposed to interact with, this should be pretty transparent for contributors.

## Context

v 0.3.95

### PR DESCRIPTION
**Pull request by stilnat**
## Summary

Currently, containers don't work on child objects. That's because Mirror doesn't allow for Network identity to be on child objects. This PR insures that all network related scripts in containers are put on the parent gameobject with a network identity.

## Pictures/Videos (optional)

Here we see two containers on the fridge, one is on the parent and one on the fridge door. They both work as expected.

image

image

## Changes to Files

- ContainerDescriptorEditor : adds some logic so that scripts are put where they belong.
- ContainerDescriptorEditor : not really related to the PR, it insures us that if any parameters of an animation is called "Open", then we get the option in editor to trigger the gameobject open animation when viewing a container. Useful for toolboxes for example.
- Small bug fix : When deleting a containerdescriptor, the container sync was removed when there was still containerdescriptors in the gameobject hierarchy. 

closes #841

### CODE REVIEW

File: Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs [hunk 1]
```diff
@@ -65,14 +65,17 @@ public override void OnInspectorGUI()
...
+        {
+            var animator = containerDescriptor.gameObject.GetComponent<Animator>();
+            if (animator != null)
             {
-                if (controllerParameter.name == "Open")
+                foreach (AnimatorControllerParameter controllerParameter in containerDescriptor.gameObject.GetComponent<Animator>().parameters)
                 {
-                    bool openWhenContainerViewed = EditorGUILayout.Toggle(new GUIContent("open when container viewed", "Set if the open animation should run when the container UI is opened"), containerDescriptor.openWhenContainerViewed);
-                    HandleOpenWhenContainerViewed(openWhenContainerViewed);
+                    if (controllerParameter.name == "Open")
```
Comment (Em3rgencyLT) [inline Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs:74]: nitpick: can reduce nesting by 1 here by inverting the if statement.

`if (controllerParameter.name != "Open") { continue; }`

File: Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs [hunk 2]
```diff
@@ -304,7 +307,15 @@ private void HandleHideItems(bool hideItems)
...
-        sp.objectReferenceValue = containerDescriptor.gameObject.AddComponent<ContainerInteractive>();
+        GameObject networkedParent = GetParentNetworkIdentity(containerDescriptor.gameObject);
+        if (networkedParent != null)
+        {
+            sp.objectReferenceValue = networkedParent.AddComponent<ContainerInteractive>();
+        }
+        else
+        {
+            sp.objectReferenceValue = containerDescriptor.gameObject.AddComponent<ContainerInteractive>();
+        }
```
Comment (Em3rgencyLT) [inline Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs:318]: Single line if/else's should be inlined imo:
```
sp.objectReferenceValue = networkedParent != null ?
networkedParent.AddComponent<ContainerInteractive>() : 
containerDescriptor.gameObject.AddComponent<ContainerInteractive>();
```

File: Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs [hunk 3]
```diff
@@ -341,15 +352,32 @@ private void AddAttached()
...
+            GameObject g = GetParentNetworkIdentity(containerDescriptor.gameObject);
+
+            if (g != null)
+            {
+                sp.objectReferenceValue = g.AddComponent<ContainerSync>();
+            }
+            else
+            {
+                sp.objectReferenceValue = containerDescriptor.gameObject.AddComponent<ContainerSync>();
+            }
```
Comment (Em3rgencyLT) [inline Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs:368]: nitpick: avoid cryptic single letter variables as per code style guidelines. Also inline the if/else.

File: Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs [hunk 4]
```diff
@@ -359,6 +387,8 @@ private void RemoveSync()
             if (containerDescriptors != null && containerDescriptors.Length == 0)
             {
                 DestroyImmediate(g.GetComponent<ContainerSync>());
+                GameObject networkedParent = GetParentNetworkIdentity(g);
+                DestroyImmediate(networkedParent.GetComponent<ContainerSync>());
```
Comment (Em3rgencyLT) [inline Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs:391]: if the GetComponent returns a null, is the DestroyImmediate ok with it? I never trust the results of a GetComponent call :D

**Review by Em3rgencyLT — CHANGES_REQUESTED** [review]

File: Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs (hunk 4 above)
Comment (stilnat) [reply Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs:375]: it's okay with it ! Tested and it handles null without any warning or errors.

File: Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs (hunk 1 above)
Comment (stilnat) [reply Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs]: done

File: Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs (hunk 2 above)
Comment (stilnat) [reply Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs]: done

File: Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs (hunk 3 above)
Comment (stilnat) [reply Assets/Engine/Inventory/Editor/ContainerDescriptorEditor.cs]: done

**Comment by cosmiccoincidence** [comment]
Push the change you made to the fridge so I too can use it as a test object.

**Review by cosmiccoincidence — APPROVED** [review]
Looks good!

I did notice 1 small error, the UI for the child container does not show the name of the container... Ill make an issue for this.

**Pull request merged**